src/main/java/org/elasticsearch/rest/action/search/RestSearchAction.java

```diff
@@ -96,8 +97,18 @@ public static SearchRequest parseSearchRequest(RestRequest request) {
             }
         }
 
+        // do not allow 'query_and_fetch' or 'dfs_query_and_fetch' search types
+        // from the REST layer. these modes are an internal optimization and should
+        // not be specified explicitly by the user.
+        String searchType = request.param("search_type");
+        if (SearchType.fromString(searchType).equals(SearchType.QUERY_AND_FETCH) ||
+                SearchType.fromString(searchType).equals(SearchType.DFS_QUERY_AND_FETCH)) {
+            throw new ElasticsearchIllegalArgumentException("Unsupported search type [" + searchType + "]");
```

Comment:
Agree that we could simply remove ```(dfs_)query_and_fetch``` from ```SearchType.fromString```. However, I feel that this would be the wrong approach. ```SearchType.fromString``` is a pure utility function that performs a single, well-defined task: Given a string it returns the corresponding enum type. Modifying the function to reflect the needs of an external, context-dependent use case strikes me as dangerous. It bakes an assumption that may not always hold into an otherwise simple, easy-to-understand function.

With respect to catching other cases such as ```more_like_this```, perhaps we might consider adding some function to ```SearchType``` to check if it is allowed to be called from public API's, e.g. ```SearchType.isInternalUseOnly```, or some such. Of course that opens the new problem that now users of the class must know about this distinction and check for it where appropriate. 

Thoughts?